File: packages/xyd-opensdk-go/src/gotype.ts
import type { TypeRef } from '@xyd-js/opensdk-core';

import { pascalCase } from './naming';

/**
 * Map an IR TypeRef to a Go type expression. Named refs resolve to their Go type
 * name in the root package; scalars map to Go builtins; arrays/maps recurse.
 */
export function goType(ref: TypeRef | undefined): string {
  if (!ref) return 'any';
  switch (ref.kind) {
    case 'scalar':
      return goScalar(ref.scalar, ref.format);
    case 'ref':
      return ref.name ? pascalCase(ref.name) : 'any';
    case 'array':
      return `[]${goType(ref.items as TypeRef | undefined)}`;
    case 'map':
      return `map[string]${goType(ref.values as TypeRef | undefined)}`;
    default:
      return 'any';
  }
}

function goScalar(scalar: string | undefined, format: string | undefined): string {
  switch (scalar) {
    case 'string':
      return 'string';
    case 'integer':
      return format === 'int32' ? 'int32' : 'int64';
    case 'number':
      return format === 'float' ? 'float32' : 'float64';
    case 'boolean':
      return 'bool';
    default:
      return 'any';
  }
}

/** Whether a TypeRef maps to a plain Go scalar (usable as a `param.Opt[T]` element). */
export function isScalarRef(ref: TypeRef | undefined): boolean {
  return ref?.kind === 'scalar';
}

/** Whether a TypeRef is a binary blob (string format=binary) — an upload payload. */
export function isBinaryRef(ref: TypeRef | undefined): boolean {
  return ref?.kind === 'scalar' && ref.scalar === 'string' && ref.format === 'binary';
}

File: crates/xyd_opensdk_go/src/gotype.rs
//! IR TypeRef → Go type — port of gotype.ts. Operates on serde_json::Value.

use serde_json::Value;

use crate::naming::pascal_case;

pub fn go_type(ref_: Option<&Value>) -> String {
    let Some(r) = ref_ else {
        return "any".to_string();
    };
    match r.get("kind").and_then(|k| k.as_str()) {
        Some("scalar") => go_scalar(
            r.get("scalar").and_then(|s| s.as_str()),
            r.get("format").and_then(|f| f.as_str()),
        ),
        Some("ref") => match r.get("name").and_then(|n| n.as_str()) {
            Some(n) if !n.is_empty() => pascal_case(n),
            _ => "any".to_string(),
        },
        Some("array") => format!("[]{}", go_type(r.get("items"))),
        Some("map") => format!("map[string]{}", go_type(r.get("values"))),
        _ => "any".to_string(),
    }
}

fn go_scalar(scalar: Option<&str>, format: Option<&str>) -> String {
    match scalar {
        Some("string") => "string",
        Some("integer") => {
            if format == Some("int32") {
                "int32"
            } else {
                "int64"
            }
        }
        Some("number") => {
            if format == Some("float") {
                "float32"
            } else {
                "float64"
            }
        }
        Some("boolean") => "bool",
        _ => "any",
    }
    .to_string()
}

pub fn is_scalar_ref(ref_: Option<&Value>) -> bool {
    ref_.and_then(|r| r.get("kind")).and_then(|k| k.as_str()) == Some("scalar")
}

pub fn is_binary_ref(ref_: Option<&Value>) -> bool {
    let Some(r) = ref_ else { return false };
    r.get("kind").and_then(|k| k.as_str()) == Some("scalar")
        && r.get("scalar").and_then(|s| s.as_str()) == Some("string")
        && r.get("format").and_then(|f| f.as_str()) == Some("binary")
}
